From 9afaa73c5ea7d0b65b3f2469196d4852cad2cabb Mon Sep 17 00:00:00 2001
From: Lingyun Zhu <lingyun.zhu@intel.com>
Date: Tue, 20 Nov 2012 19:24:35 +0800
Subject: [PORT FROM MAIN] Graphics:Add error check when create GraphicBuffer

BZ: 50141

Port JB 4.1 patch 60108 to JB 4.2.1

Change-Id: I608d9a9028dec9562c2b58a7b9b6a370404129da
Orig-Change-Id: I9b251082e82be1a0f100c395359f4a485a1c1e69
Signed-off-by: Lingyun Zhu <lingyun.zhu@intel.com>
Reviewed-on: http://android.intel.com:8080/74864
Reviewed-by: Yao, Yong <yong.yao@intel.com>
Tested-by: Tong, BoX <box.tong@intel.com>
Reviewed-by: Tong, BoX <box.tong@intel.com>
Reviewed-by: buildbot <buildbot@intel.com>
Tested-by: buildbot <buildbot@intel.com>
---
 libs/gui/IGraphicBufferProducer.cpp |    6 +++++-
 libs/ui/GraphicBuffer.cpp           |    2 ++
 2 files changed, 7 insertions(+), 1 deletions(-)

diff --git a/libs/gui/IGraphicBufferProducer.cpp b/libs/gui/IGraphicBufferProducer.cpp
index 63d7628..c335439 100644
--- a/libs/gui/IGraphicBufferProducer.cpp
+++ b/libs/gui/IGraphicBufferProducer.cpp
@@ -62,7 +62,11 @@ public:
         bool nonNull = reply.readInt32();
         if (nonNull) {
             *buf = new GraphicBuffer();
-            reply.read(**buf);
+            result = reply.read(**buf);
+            if(result != NO_ERROR) {
+                (*buf).clear();
+                return result;
+            }
         }
         result = reply.readInt32();
         return result;
diff --git a/libs/ui/GraphicBuffer.cpp b/libs/ui/GraphicBuffer.cpp
index b9cab85..c084cd0 100644
--- a/libs/ui/GraphicBuffer.cpp
+++ b/libs/ui/GraphicBuffer.cpp
@@ -260,6 +260,8 @@ status_t GraphicBuffer::unflatten(void const* buffer, size_t size,
     if (handle != 0) {
         status_t err = mBufferMapper.registerBuffer(handle);
         if (err != NO_ERROR) {
+            width = height = stride = format = usage = 0;
+            handle = NULL;
             ALOGE("unflatten: registerBuffer failed: %s (%d)",
                     strerror(-err), err);
             return err;
-- 
1.7.4.1

